lightning-invoice/src/utils.rs

```diff
@@ -627,8 +683,49 @@ fn filter_channels<L: Deref>(
 
 			include_channel
 		})
-		.map(route_hint_from_channel)
-		.collect::<Vec<RouteHint>>()
+		.collect::<Vec<ChannelDetails>>();
+
+		eligible_channels.sort_unstable_by(|a, b| {
+			if online_min_capacity_channel_exists {
+				a.inbound_capacity_msat.cmp(&b.inbound_capacity_msat)
+			} else {
+				b.inbound_capacity_msat.cmp(&a.inbound_capacity_msat)
+			}});
+		eligible_channels.into_iter().take(3).map(route_hint_from_channel).collect::<Vec<RouteHint>>()
+}
+
+// prefer_current_channel chooses a channel to use for route hints between a currently selected and candidate 
+// channel based on the inbound capacity of each channel and the minimum inbound capacity requested for the hints, 
+// returning true if the current channel should be preferred over the candidate channel.
+// * If no minimum amount is requested, the channel with the most inbound is chose to maximize the chances that a
+//   payment of any size will succeed.
+// * If we have channels with inbound above our minimum requested inbound (plus a scaling factor to allow some leeway)
+//   then we choose the lowest inbound channel with above this amount. If we have sufficient inbound channels, we don't 
+//   want to deplete our larger channels with small payments (the off-chain version of "grinding our change").
+// * If no channel above our minimum amount exists, then we just prefer the channel with the most inbound to give 
+//   payments the best chance of succeeding in multiple parts.
+fn prefer_current_channel(min_inbound_capacity_msat: Option<u64>, scale_factor: f64, current_channel: u64,
+	candidate_channel: u64) -> bool {
+
+	// If no min amount is given for the hints, err of the side of caution and choose the largest channel inbound to 
+	// maximize chances of any payment succeeding.
+	if min_inbound_capacity_msat.is_none() {
```

Comment:
Not sure how to deal with `None` min inbound amounts - do we want to prefer larger channels because we don't know how much they'll be (which is what this does) or smaller channels because  0 amount invoices are likely small tips?